services/route_manager/src/main/java/com/futurewei/alcor/route/controller/NeutronRouterController.java

```diff
@@ -0,0 +1,375 @@
+        if (routerExtraAttribute == null) {
+            return new ResponseId();
+        }
+        this.routerExtraAttributeDatabaseService.deleteRouterExtraAttribute(routerExtraAttribute.getId());
+
+        return new ResponseId(routerid);
+
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/routers/{routerid}/add_router_interface"})
+    @DurationStatistics
+    public RouterInterfaceResponse addInterfaceToNeutronRouter(@PathVariable String projectid, @PathVariable String routerid, @RequestBody RouterInterfaceRequest resource) throws Exception {
+
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(routerid);
+        RestPreconditionsUtil.verifyResourceFound(projectid);
+
+        String portId = resource.getPortId();
+        String subnetId = resource.getSubnetId();
+
+        RouterInterfaceResponse routerInterfaceResponse = this.neutronRouterService.addAnInterfaceToNeutronRouter(projectid, portId, subnetId, routerid);
+
+        return routerInterfaceResponse;
+
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/routers/{routerid}/remove_router_interface"})
+    @DurationStatistics
+    public RouterInterfaceResponse removeInterfaceToNeutronRouter(@PathVariable String projectid, @PathVariable String routerid, @RequestBody RouterInterfaceRequest resource) throws Exception {
+
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(routerid);
+        RestPreconditionsUtil.verifyResourceFound(projectid);
+
+        String portId = resource.getPortId();
+        String subnetId = resource.getSubnetId();
+
+        RouterInterfaceResponse routerInterfaceResponse = this.neutronRouterService.removeAnInterfaceToNeutronRouter(projectid, portId, subnetId, routerid);
+
+        return routerInterfaceResponse;
+
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/routers/{routerid}/add_extra_routes"})
+    @DurationStatistics
+    public RoutesToNeutronWebResponse addRoutesToNeutronRouter(@PathVariable String projectid, @PathVariable String routerid, @RequestBody RoutesToNeutronWebRequest resource) throws Exception {
+
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+        RestPreconditionsUtil.verifyParameterNotNullorEmpty(routerid);
+        RestPreconditionsUtil.verifyResourceFound(projectid);
+
+        RoutesToNeutronRouterRequestObject router = resource.getRouter();
+        if (router == null) {
+            return new RoutesToNeutronWebResponse();
+        }
+
+        RoutesToNeutronWebResponse routesToNeutronWebResponse = this.neutronRouterService.addRoutesToNeutronRouter(routerid, router);
+
```

Comment:
@kevin-zhonghao we need to add a function call here for l3-neighbors-updating. It should be the function call to PM same as the previous one /project/{project_id}/update-l3-neighbors/{new_subnet_id}, but the detail contract is not confirm yet.  We can wait until the contract is confirmed.

Reply:
Sure ~ left the comments already